src/main/java/edu/cornell/kfs/concur/service/impl/ConcurAccessTokenServiceImpl.java

```diff
@@ -48,25 +53,49 @@
     private String concurRequestAccessTokenURL;
     private String concurRefreshAccessTokenURL;
     private String concurRevokeAccessTokenURL;
-    protected WebServiceCredentialService webServiceCredentialService;
+    protected AwsSecretService awsSecretService;
     protected ConcurBatchUtilityService concurBatchUtilityService;
 
+    protected ConcurStaticConfig getStaticConfig() {
+        return getConcurPojoFromAws(ConcurAwsKeyNames.STATIC_CONFIG, ConcurStaticConfig.class);
```

Comment:
I could add instances variables instead of static variables to cache the tokens, but it would be helpful to have some way to manually or periodically flush the cache, in case we have a need to manually change the token through the UI or the Secrets Manager. Perhaps I could add a service to the bus (or set up a cache with EHCache) that could clear out or re-retrieve the tokens via the KFS Concur screen, and that would at least take care of the KFS runtime instances. As for KFS Batch, we could just request a batch container restart if we need it to pick up a manual token change.

Reply:
That is a good point about refreshing on demand.  What if the cache value is "scoped" to a job run?  Meaning when the job run it pulls the secrets from AWS, then using the in memory values for the rest of the job run.  Then start fresh again.

Reply:
It may not be practical to "scope" the tokens on a per-job-run basis, due to the complexities it would introduce for determining when and how the "scope" starts or ends.

After some further thought, there could be another approach that would work around the difficulties with the EHCache distributed caching. I could set up a local in-memory cache of AWS secrets, but they would calculate a special string from the secret value (such as a checksum or timestamp), and then store the special value in the database (such as through the Web Service Credential service or a brand new service). Later, when retrieving a secret value from the cache, it can compare the locally-stored timestamp/checksum against the DB-stored one, to determine whether the secret value needs to be re-retrieved from AWS.

Please let me know your thoughts about this, or if you think this is introducing an unnecessary amount of extra caching code.

Reply:
I think we are trending towards too complicated with that approach.  I was looking towards reducing the number of times we had to retrieve a secret from AWS, but if we over complicate the process, extra time spent supporting it would over shadow savings.

Reply:
In that case, I'm not aware of an appropriate alternative, unless we want a per-instance cache that just expires its elements after a certain amount of time. If there's enough spare time left at the end of tomorrow's MOPS meeting, we could check with the rest of the team to get their thoughts on what (if any) AWS secrets caching would be best. (We might also be able to use that time to discuss your other concern about the example-jaxrs-server unit test.)